exercises/src/main/scala/fpinscala/parallelism/Nonblocking.scala

```diff
@@ -16,28 +18,34 @@ object Nonblocking {
     def run[A](es: ExecutorService)(p: Par[A]): A = {
       val ref = new java.util.concurrent.atomic.AtomicReference[A] // A mutable, threadsafe reference, to use for storing the result
       val latch = new CountDownLatch(1) // A latch which, when decremented, implies that `ref` has the result
-      p(es) { a => ref.set(a); latch.countDown } // Asynchronously set the result, and decrement the latch
+
+      // Asynchronously set the result, and decrement the latch
+      p(es) { a =>
+        ref.set(a)
+        latch.countDown
+      } { kErr =>
+        latch.countDown
+        throw kErr
+      }
+
       latch.await // Block until the `latch.countDown` is invoked asynchronously
       ref.get // Once we've passed the latch, we know `ref` has been set, and return its value
     }
 
     def unit[A](a: A): Par[A] =
       es => new Future[A] {
-        def apply(cb: A => Unit): Unit =
-          cb(a)
+        def apply(k: A => Unit)(kErr: Throwable => Unit): Unit =
+          try { k(a) }
+          catch { case err: Throwable => kErr(err) }
       }
 
     /** A non-strict version of `unit` */
-    def delay[A](a: => A): Par[A] =
-      es => new Future[A] {
-        def apply(cb: A => Unit): Unit =
-          cb(a)
-      }
+    def delay[A](a: => A): Par[A] = unit(a)
```

Comment:
@gpoirier `unit` is indeed strict. You'll notice in da6b7583, `unit` and `delay` had the exact same implementation--the only difference was that `delay` took its argument `a` by name and `unit` took it by value. So we can simplify here.